### PR TITLE
Feature/issue28 test kpl properties

### PR DESCRIPTION
**Pull request by markglh**
Fixes: 
https://github.com/WW-Digital/reactive-kinesis/issues/28 and https://github.com/WW-Digital/reactive-kinesis/issues/10

* Refactored how the properties are loaded to share commonality between the actor and Scala implementations. This simplifies the Scala implementation and makes it more robust. It also allows more control over the underlying KPL properties should someone want to not use the Typesafe config approach.
* Removed the pointless trait in the Producer and renamed `KinesisProducerKPL` to `KinesisProducer` - that makes this a `very small` breaking change.
* The Producer properties are now automatically tested reflectively against the underlying KPL implementation, so our tests will fail automatically when new ones are added or existing ones changed. I'll do the same for the Consumer in a different PR.
* Tidied a few test exceptions and (hopefully) fixed an intermittent test
* Updated the readme to show the tweaked usage of the Scala implementation.

Raised a PR on the AWS KPL to make this a bit better in the future, it's really frustrating how inconsistent the KPL and KCL are in this regard.

### CODE REVIEW

File: src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+        // TODO we don't yet support setting of credentials providers via config due to KPL limitations
+        // TODO see issue #29 : https://github.com/WW-Digital/reactive-kinesis/issues/29
+        .filterNot(_.toLowerCase.contains("credentialsprovider"))
+
+      configKeys foreach { configKey =>
+        val field =
+          kplLibConfiguration.getClass.getDeclaredField(configKey.head.toLower + configKey.tail)
+        field.setAccessible(true)
+
+        println(s"${field.getName}=${field.get(kplLibConfiguration)}")
```
Comment (j-potts) [inline src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: println

File: src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+      val kplConfig = kinesisConfig2.getConfig("testProducer.kpl")
+
+      val kplLibConfiguration = producerConf.kplLibConfiguration
+      kplLibConfiguration.isAggregationEnabled should be(false)
+
+      //We're dealing with Java classes so using Java reflection is cleaner here
+      //Start with the setters to prevent picking up all the unrelated private fields, stripping the "set"
+      val configKeys = kplLibConfiguration.getClass.getDeclaredMethods
+        .filter(_.getName.startsWith("set"))
+        .map(_.getName.drop(3))
```
Comment (j-potts) [inline src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: Yeesh, this seems tricky. We should add another issue to find a way to work with this, without the reflection if possible  :/

File: src/main/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConf.scala [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            producerName: String,
+            credentialsProvider: Option[AWSCredentialsProvider] = None): ProducerConf = {
+
+    val producerConfig = kinesisConfig
+      .getConfig(producerName)
+      .withFallback(kinesisConfig.getConfig("default-producer"))
+
+    val streamName = producerConfig.getString("stream-name")
+    require(!streamName.isEmpty,
+            "A valid stream name must be provided to start the Kinesis Producer")
```
Comment (j-potts) [inline src/main/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConf.scala]: Maybe instead, reword this as something like `Config field "stream-name" missing, a value must be provided!`

This way the dev knows the exact field he's missing, and where he's missing it from.

File: src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala (hunk 2 above)
Comment (markglh) [reply src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: not possible without reflection. We're introspecting the java class to make sure we're covering it all. The alternative is to manually specify each property in the spec, but that isn't future proof.
I use setters because I only care about stuff we can actually set.

Comment (markglh) [reply src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: It's pretty solid, if not pretty

Comment (j-potts) [reply src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: Yea, just a bit ugly, but if it's a hard constraint it'll have to do

Comment (markglh) [reply src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: scala reflection is even uglier imho :D

Comment (markglh) [reply src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: you could just do getFields instead, but then you'd have to account for other private fields (like loggers)

Comment (agaro1121) [reply src/test/scala/com/weightwatchers/reactive/kinesis/producer/ProducerConfSpec.scala]: it's a test class so i'm not sure if it's too bad
